### PR TITLE
Scala native support

### PR DESCRIPTION
**Pull request by Baccata**
Adds Scala Native support (for Scala 3 only, to avoid the matrix increasing in size too much for something that is quite bleeding edge). 

The supported modules are : 

* smithy4s-core
* smithy4s-json 
* smithy4s-dynamic 

This PR also amends the dynamic module slightly to make it depend on the smithy-model library (on JVM only), facilitating the loading of dynamic services/schemas directly from a smithy Model. This makes tests a little bit easier to write and reduces the need for sharing some test-utilities across modules, which was leading to a cyclic module dependency that I was too lazy to solve via yet another indirection. 

Additionally, because smithy4s-dynamic is built on native now, I had to remove the dependency to CE/weaver in its test. I wanted to keep the diff smallish, so I ended up making a dummy IO construct (totally not pure) to reduce the noise a little. When CE builds on native, we'll switch back.

### CODE REVIEW

File: modules/dynamic/test/src-jvm/smithy4s/dynamic/Http4sDynamicPizzaClientSpec.scala [hunk 1]
```diff
@@ -1,48 +0,0 @@
...
-            JsonIOProtocol
-              .fromJsonIO[PizzaAdminServiceGen, PizzaAdminServiceOperation](
-                JsonIOProtocol.toJsonIO(dynamicClient)(dsi.service)
-              )
-
-          }
-      }
-}
-
-object Http4sDynamicPizzaClientSpec extends smithy4s.tests.PizzaClientSpec {
```
Comment (Baccata) [inline modules/dynamic/test/src-jvm/smithy4s/dynamic/Http4sDynamicPizzaClientSpec.scala:40]: moved this one to the http4s module.

File: modules/dynamic/test/src/smithy4s/dynamic/DummyIO.scala [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s.dynamic
+import smithy4s.tests._
+
+object DummyIO {
```
Comment (Baccata) [inline modules/dynamic/test/src/smithy4s/dynamic/DummyIO.scala:20]: They call me Haxxor, destroyer of common-sense. 

Seriously though, it was easier to do that than turn everything into a `scala.util.Try`

File: modules/dynamic/test/src/smithy4s/dynamic/KVStoreImpl.scala [hunk 3]
```diff
@@ -18,29 +18,25 @@ package smithy4s
...
+import DummyIO._
 
-class KVStoreImpl(ref: Ref[IO, Map[String, String]]) extends KVStore[IO] {
-  def delete(key: String): IO[Unit] = ref
-    .modify[Either[Throwable, Unit]] { map =>
-      map.get(key) match {
-        case None        => (map, Left(KeyNotFoundError(key)))
-        case Some(value) => (map - value, Right(()))
-      }
+class KVStoreImpl(map: scala.collection.mutable.Map[String, String])
```
Comment (Baccata) [inline modules/dynamic/test/src/smithy4s/dynamic/KVStoreImpl.scala:23]: sue me.

File: build.sbt [hunk 4]
```diff
@@ -423,10 +432,14 @@ lazy val dynamic = projectMatrix
   .jvmPlatform(
     allJvmScalaVersions,
     jvmDimSettings ++ Seq(
-      libraryDependencies += Dependencies.Smithy.model % Test
+      Test / fork := true,
+      // Forwarding cwd to tests
```
Comment (daddykotex) [inline build.sbt:436]: the comment is wrong in the sense that this is now the cwd

File: modules/dynamic/test/src/smithy4s/dynamic/DummyIO.scala (hunk 2 above)
Comment (daddykotex) [reply modules/dynamic/test/src/smithy4s/dynamic/DummyIO.scala:20]: ahhaha that's clever

**Comment by daddykotex** [comment]
Just asking, would you accept a PR where I update the copyright everywhere, that we would merge to main (and eventually here) so that we reduce the noise in this PR ?

**Comment by Baccata** [comment]
I would certainly

File: build.sbt (hunk 4 above)
Comment (Baccata) [reply build.sbt:436]: I'm not sure I follow. Care to elaborate ?

**Comment by daddykotex** [comment]
> I would certainly

see https://github.com/disneystreaming/smithy4s/pull/292

File: build.sbt [hunk 5]
```diff
@@ -96,6 +97,20 @@ lazy val docs =
     )
     .settings(Smithy4sPlugin.doNotPublishArtifact)
 
+val munitDeps = Def.setting {
+  if (virtualAxes.value.contains(VirtualAxis.native)) {
+    Seq(
+      Dependencies.MunitMilestone.core.value % Test,
```
Comment (daddykotex) [inline build.sbt:103]: what's the reason why we can't switch to the Milestone all together

File: modules/core/src-jvm-native/smithy4s/Timestamp.scala [hunk 6]
```diff
@@ -17,11 +17,11 @@
 package smithy4s
 
 import smithy.api.TimestampFormat
-import smithy4s.Timestamp._
-import java.time._
+// import smithy4s.Timestamp._
```
Comment (daddykotex) [inline modules/core/src-jvm-native/smithy4s/Timestamp.scala:20]: can we get rid of the comment here?

File: project/Smithy4sPlugin.scala [hunk 7]
```diff
@@ -445,15 +447,21 @@ object Smithy4sPlugin extends AutoPlugin {
           val platformAxis =
             if (projectId.endsWith(jsSuffix)) {
               projectId = projectId.dropRight(jsSuffix.length)
-
               "js"
-            } else "jvm"
+            } else if (projectId.endsWith(nativeSuffix)) {
+              projectId = projectId.dropRight(nativeSuffix.length)
```
Comment (daddykotex) [inline project/Smithy4sPlugin.scala:452]: gotta love the mutation :D

**Review by daddykotex — APPROVED** [review]
I'm ok with this PR.

I had one question regarding Munit but it's not a blocker and could be resolved w/ a comment I believe

**Pull request merged**